ostd/src/cpu/local/cell.rs

```diff
@@ -0,0 +1,146 @@
+///     // You can avoid this by disabling interrupts (and preemption, if needed).
+///     println!("BAR VAL: {:?}", BAR.load());
+///
+///     let _irq_guard = ostd::trap::disable_local_irq();
+///     println!("1st FOO VAL: {:?}", FOO.load());
+///     // No suprises here, the two accesses must result in the same value.
+///     println!("2nd FOO VAL: {:?}", FOO.load());
+/// }
+/// ```
+macro_rules! cpu_local_cell {
+    ($( $(#[$attr:meta])* $vis:vis static $name:ident: $t:ty = $init:expr; )*) => {
+        $(
+            #[link_section = ".cpu_local"]
+            $(#[$attr])* $vis static $name: $crate::cpu::local::CpuLocalCell<$t> = {
+                let val = $init;
+                // SAFETY: The CPU local variable instantiated is statically
+                // stored in the special `.cpu_local` section.
+                unsafe {
+                    $crate::cpu::local::CpuLocalCell::__new(val)
+                }
+            };
+        )*
+    };
+}
+
+pub(crate) use cpu_local_cell;
+
+/// Inner mutable CPU-local objects.
+///
+/// CPU-local cell objects are only accessible from the current CPU. When
+/// accessing an underlying object using the same `CpuLocalCell` instance, the
+/// actually accessed object is always on the current CPU. So in a preemptive
+/// kernel task, the operated object may change if interrupts are enabled.
+///
+/// The inner mutability is provided by single instruction operations, and the
+/// CPU-local cell objects will not ever be shared between CPUs. So it is safe
+/// to modify the inner value without any locks.
+///
+/// You should only create the CPU-local cell object using the macro
+/// [`cpu_local_cell!`].
+///
+/// For the difference between [`super::CpuLocal`] and [`CpuLocalCell`], see
+/// [`super`].
+pub struct CpuLocalCell<T: 'static>(UnsafeCell<T>);
+
+#[cfg(debug_assertions)]
+use core::sync::atomic::Ordering;
+
+#[cfg(debug_assertions)]
+use super::IS_INITIALIZED;
+
+impl<T: 'static> CpuLocalCell<T> {
+    /// Initialize a CPU-local object.
+    ///
+    /// Please do not call this function directly. Instead, use the
+    /// `cpu_local!` macro.
+    ///
+    /// # Safety
+    ///
+    /// The caller should ensure that the object initialized by this
+    /// function resides in the `.cpu_local` section. Otherwise the
+    /// behavior is undefined.
+    #[doc(hidden)]
+    pub const unsafe fn __new(val: T) -> Self {
+        Self(UnsafeCell::new(val))
+    }
+
+    /// Get access to the underlying value through a raw pointer.
+    ///
+    /// This function calculates the virtual address of the CPU-local object
+    /// based on the CPU-local base address and the offset in the BSP.
+    ///
+    /// # Safety
+    ///
+    /// The caller should ensure that within the entire execution of this
+    /// function, no interrupt or preemption can occur. Otherwise, the
+    /// returned pointer may points to the variable in another CPU.
+    pub unsafe fn get(&'static self) -> *mut T {
```

Comment:
This is in accordance with the [`UnsafeCell::get`](https://doc.rust-lang.org/std/cell/struct.UnsafeCell.html#method.get) method, so do the naming. Interestingly `UnsafeCell::get` is safe but it is unsafe due to the preemption problem.

It is used in `task/processor.rs` to create a CPU-local writable place for the assembly to store the bootstrap `TaskContext`.

Reply:
> This is in accordance with the [UnsafeCell::get](https://doc.rust-lang.org/std/cell/struct.UnsafeCell.html#method.get) method

For a type `U` that contains a value of `T`,  a `get` method is expected to be the most natural getter to access `T`. When we name a method `get`, we are essentially inviting the user to use it because the name gives users an impression of cheap and unharmful.

* `UnsafeCell<T>::get` naturally returns `*mut T`, because `UnsafeCell` is essentially a mutable location.
* `Cell<T>::get` naturally returns `T`.
* `LocalKey<T: Copy>::get` naturally returns `T`.

Now let's consider what is the natural getter for `CpuLocalCell<T>`. Pause here. And think of an answer for yourself.

My answer is for `CpuLocalCell::get` to return `T: Copy`, same as `LocalKey<T: Copy>`. As a user, I would not want a `get` method to return a pointer, which cannot be used safely. So the getter for the pointer should be named `as_ptr` or `as_mut_ptr`.

This is not to say that we should actually implement a `get` method that returns `T`. This is because such a `get` method (the current `load` method) is tricky to implement and sometimes may be more expensive than it appears (disabling IRQs). In short, `CpuLocalCell` should not have a `get` method.